design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,223 @@
+# Monitoring code refactor
+
+**Authors:** João Vilaça <jvilaca@redhat.com>
+
+## What
+
+This design document is proposing a code refactor for the monitoring logic in
+all KubeVirt components. The goal is to have a centralized place for the
+creation of monitoring resources, a consistent monitoring package between all
+projects, a code struct that is easy to maintain and evolve, while moving closer
+to the Kubernetes metric implementation style.
+
+## Why
+
+It’s important to refactor the monitoring code to separate monitoring logic from
+business logic. It will make the codebase more modular, readable, and
+maintainable. It will reduce complexity, reduce the risk of introducing errors,
+and allow us to test each component independently.
+
+Right now, every change regarding monitoring needs to happen at many different
+places. For example, when adding the linter for metrics names, it was necessary
+to go to all KubeVirt components repositories and implement the linter
+validation in each. This is demanding, very time-consuming, requires going
+through multiple duplicated code reviews, uses many CI resources, and makes it
+mentally harder to decide to change anything.
+
+### Pitfalls of the current solution
+
+We are facing multiple challenges with the current solution:
+
+- It is hard to keep a mental model of how monitoring is implemented in all the
+different repositories and is difficult to maintain all of them in sync with new
+features
+
+- Hard to add new features to metrics and alerts, for example, because of how
+metrics are being created it is hard to add a new metric stability field to all
+of them since they are independently created in many different places
```

Comment:
I don't understand this statement. Why is it hard? 
Why can't we add the same logic to each operator?

Reply:
I think the emphasis here is that :
```
Since metrics are being created independently in many different places, its hard to get the full list of metrics that the operator is exposing and this is causing issues. For example:
- We miss metrics when doing automatic documentation
- We don't test all metrics with the needed tests
- We can apply additional feature to the metrics that we are not aware of
```